torch_geometric/nn/models/rbcd_attack.py

```diff
@@ -0,0 +1,760 @@
+
+# (predictions, labels, ids/mask) -> Tensor with one element
+LOSS_TYPE = Callable[[Tensor, Tensor, Optional[Tensor]], Tensor]
+
+
+class PRBCDAttack(torch.nn.Module):
+    r"""Projected Randomized Block Coordinate Descent (PRBCD) adversarial
+    attack from the `Robustness of Graph Neural Networks at Scale
+    <https://www.cs.cit.tum.de/daml/robustness-of-gnns-at-scale>`_ paper.
+
+    This attack uses an efficient gradient based approach that (during the
+    attack) relaxes the discrete entries in the adjacency matrix
+    :math:`\{0, 1\}` to :math:`[0, 1]` and solely perturbs the adjacency matrix
+    (no feature perturbations). Thus, this attack supports all models that can
+    handle weighted graphs that are differentiable w.r.t. these edge weights.
+    For non-differentiable models you might be able to e.g. use the gumble
+    softmax trick.
+
+    The memory overhead is driven by the additional edges (at most
+    :attr:`block_size`). For scalability reasons, the block is drawn with
+    replacement and then the index is made unique. Thus, the actual block size
+    is typically slightly smaller than specified.
+
+    This attack can be used for both global and local attacks as well as
+    test-time attacks (evasion) and training-time attacks (poisoning). Please
+    see the provided examples.
+
+    This attack is designed with a focus on node- or graph-classification,
+    however, to adapt to other tasks you most likely only need to provide an
+    appropriate loss and model. However, we currently do not support batching
+    out of the box (sampling needs to be adapted).
+
+    .. note::
+        For examples of using the PRBCD Attack, see
+        `examples/rbcd_attack.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/examples/
+        rbcd_attack.py>`_
+        for a test time attack (evasion) or `examples/rbcd_attack_poisoning.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/examples/
+        rbcd_attack_poisoning.py>`_
+        for a training time (poisoning) attack.
+
+    Args:
+        model (torch.nn.Module): The GNN module to assess.
+        block_size (int, optional): Number of randomly selected elements in the
+            adjacency matrix to consider. (default: :obj:`250_000`)
+        epochs (int, optional): Number of epochs (aborts early if
+            :obj:`mode='greedy'` and budget is satisfied) (default: :obj:`125`)
+        epochs_resampling (int, optional): Number of epochs to resample the
+            random block. (default: obj:`100`)
+        loss (str or Callable, optional): A loss to quantify the "strength" of
+            an attack. Note that this function must match the output format of
+            :attr:`model`. By default, it is assumed that the task is
+            classification, that the model returns raw predictions (i.e. no
+            output activation) or uses :obj:`logsoftmax`, and that the number
```

Comment:
Does it not matter whether we return raw or log-probs?

Reply:
Applying `log_softmax` multiple consecutive times does not make a difference (except for a slightly higher memory footprint).